### ISSUE TITLE
addAttributeToFilter in Product List is no effect

### ISSUE THREAD

**Issue by forgeonline** [magento/magento2#4405]
## Steps to reproduce
1. Create a Product attribute (Text Field) named groupcatalogavailability which will programmed to add Customer Groups to given product in backend. All works fine (Use text values for testing)
2. write a event hook for catalog_block_product_list_collection
3. Apply addAttributeToFilter to observer to filter listing with testfield
## Expected result
1. Should filter result base on the value in groupcatalogavailability Attribute and customer group
## Actual result
1. Not filtering as expected
2. It ignores (at_groupcatalogavailability.value  IS NOT NULL) part of the query
3. However when I remove <code> INNER JOIN `ero_search_tmp_572a79b72d2122_56564892` AS `search_result` ON e.entity_id = search_result.entity_id</code> part and run the query, it filter out base on  (at_groupcatalogavailability.value IS NOT NULL) 

My Observer

``` php
public function execute(\Magento\Framework\Event\Observer $observer)
{
$productCollection = $observer->getEvent()->getCollection();
        if ( !$this->_customerSession->isLoggedIn() ) {
            $customerGroupId = 0;
        } else {
            $customerGroupId = $this->_customerSession->getCustomerGroupId();
        }
        if( $customerGroupId==0 ){
            $productCollection->addAttributeToFilter('groupcatalogavailability', ['null' => true]);
        }
        return $this;
    }
```

SQL output 

``` sql
SELECT `e`.*, `cat_index`.`position` AS `cat_index_position`, `price_index`.`price`, 
`price_index`.`tax_class_id`, `price_index`.`final_price`, IF(price_index.tier_price IS NOT NULL, 
LEAST(price_index.min_price, price_index.tier_price), price_index.min_price) AS `minimal_price`, 
`price_index`.`min_price`, `price_index`.`max_price`, `price_index`.`tier_price`, 
IF(at_sort.value_id > 0, at_sort.value, at_sort_default.value) AS `sort`, 
`at_groupcatalogavailability`.`value` AS `groupcatalogavailability` FROM 
`ero_catalog_product_entity` AS `e` INNER JOIN `ero_catalog_category_product_index` AS
 `cat_index` ON cat_index.product_id=e.entity_id AND cat_index.store_id=1 AND cat_index.visibility 
IN(2, 4) AND cat_index.category_id='9' INNER JOIN `ero_cataloginventory_stock_status` AS 
`stock_status_index` ON e.entity_id = stock_status_index.product_id AND 
stock_status_index.website_id = '1' AND stock_status_index.stock_id = 1 
INNER JOIN `ero_catalog_product_index_price` AS `price_index` ON price_index.entity_id = e.entity_id AND price_index.website_id = '1' AND price_index.customer_group_id = 0 
INNER JOIN `ero_search_tmp_572a79b72d2122_56564892` AS `search_result` ON e.entity_id = search_result.entity_id 
LEFT JOIN `ero_catalog_product_entity_varchar` AS `at_sort_default` ON (`at_sort_default`.`entity_id` = `e`.`entity_id`) AND (`at_sort_default`.`attribute_id` = '136') AND `at_sort_default`.`store_id` = 0 LEFT JOIN `ero_catalog_product_entity_varchar` AS `at_sort` ON (`at_sort`.`entity_id` = `e`.`entity_id`) AND (`at_sort`.`attribute_id` = '136') AND (`at_sort`.`store_id` = 1) 
INNER JOIN `ero_catalog_product_entity_varchar` AS `at_groupcatalogavailability` ON (`at_groupcatalogavailability`.`entity_id` = `e`.`entity_id`) 
AND (`at_groupcatalogavailability`.`attribute_id` = '144') 
AND (`at_groupcatalogavailability`.`store_id` = 0) 
WHERE (stock_status_index.stock_status=1) AND (at_groupcatalogavailability.value  IS NOT NULL)
 ORDER BY `sort` ASC LIMIT 15
```

**Comment by andimov** [comment]
@forgeonline 
Thank you for reporting!
Please provide the used Magento version.

**Comment by forgeonline** [comment]
Magento 2.0.7

**Comment by pyyick** [comment]
Any updates on it? I encounter the same problem on Magento 2.1.2

**Comment by magento-engcom-team** [comment]
@forgeonline, thank you for your report.
The issue is already fixed in 2.2.0
